### PR TITLE
add bancor views

### PR DESCRIPTION
**Pull request by milkyklim**
I've checked that:

* [x] the folder name matches the schema name
* [x] the schema name exists in Dune
* [x] the view name starts with `view_`
* [x] the filename matches the defined view and ends with .sql
* [x] each file has only one view defined  
* [x] column names are `lowercase_snake_cased`

### CODE REVIEW

**Comment by milkyklim** [comment]
@mewwts 
- Can you check that the views are alright?  
- Should I add missing tokens that are used by bancor to `erc20.tokens` as a separate PR?

**Comment by hagaetc** [comment]
Hey @milkyklim - nice work as always:) 

I think it best if you join or union all the conversion events together - they are different tables in Dune because they have different ABIs for different versions, but it makes sense to query them as one table. 

Yes, please add missing tokens in `erc20.tokens` in a separate PR.

**Comment by milkyklim** [comment]
Will update, thanks for review!

**Comment by milkyklim** [comment]
@hagaetc I've converted multiple events into one [view_convert.sql](https://github.com/duneanalytics/abstractions/blob/513231e634ac94e7c98fa0835ae09c49f9a31688/schema/bancor/view_convert.sql)

However, I am not 100% sure that all events are actually in Dune DBs, e.g. for this [transaction](https://etherscan.io/tx/0x00001382ccbbd5fd6f5e66dbe7eb4c8405c447e2fe41e4b50e7ece6097997d52) includes 3 `Conversion`s and they are in DB, while this [transaction](https://etherscan.io/tx/0x00001a7af2bf99e5117d0403b8588bad6cc4f2f9298dafe6c6e805467aae89c0) also has 3 events but DB shows only 2.

**Comment by mewwts** [comment]
FYI I'm just waiting for @hagaetc and the bancor team to confirm!

**Comment by hagaetc** [comment]
Hey @milkyklim - sorry for the delay here. It is rather hard to get all the Bancor data but all recent conversions should be included in `bancor."BancorConverterWithFee_evt_Conversion"` - the missing event you mentioned is from 2 years ago. 

I am not familiar with the details of Bancor, but I assume the BNT<>USDB cases are solved correctly. Not that you can make a PR to `erc20.tokens` as well if we lack symbols for some tokens.

All over I think this looks good and can be merged :)

**Comment by milkyklim** [comment]
@hagaetc, you seem to be right about `bancor."BancorConverterWithFee_evt_Conversion"` but I wanted to make views as full as possible, specifically:

- `bancor."BancorConverter_evt_Conversion"` is fully contained in `bancor."BancorConverterWithFee_evt_Conversion"`
- `bancor."BancorNetwork_v1_evt_Conversion"` partially overlaps with `bancor."BancorConverterWithFee_evt_Conversion"`
- `bancor."BancorLegacyConverter_evt_Conversion"` partially overlaps with `bancor."BancorConverterWithFee_evt_Conversion"`

Therefore, I've added this huge union in [view_convert.sql](https://github.com/duneanalytics/abstractions/blob/57a79031f10d241fef57988cc8d5ace8183bb95f/schema/bancor/view_convert.sql). 

`BNT<>USDB` exception is there cause `BNT<>USDB` pool won't be processed correctly. :) 

I've fixed the issue with the semicolons, see above.

**Comment by hagaetc** [comment]
✅✅Sounds good!

**Comment by milkyklim** [comment]
@mewwts need some help on your side. 

Can you double check that following tables and values are parsed correctly:
- bancor."BancorConverterWithFee_evt_Conversion", `trader` column is a contract address but should not it be equal to transaction sender?
- bancor."BancorLegacyConverter_evt_Conversion", `trader` same.
- bancor."BancorConverter_evt_Conversion", `trader` same.

**Comment by mewwts** [comment]
Happy to help, can you link me some examples on telegram?

**Comment by milkyklim** [comment]
@hagaetc these are the final changes from my side.

File: schema/bancor/view_convert.sql [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+   FROM erc20.tokens
+   UNION SELECT *
+   FROM bancor.view_smart_tokens) t2 ON s.target_token_address = t2.contract_address
+LEFT JOIN prices.usd p2 ON p2.minute = date_trunc('minute', s.block_time)
+    AND p2.symbol = t2.symbol
+;
+
+CREATE INDEX IF NOT EXISTS view_convert_idx_1 ON bancor.view_convert (source_token_address) INCLUDE (source_token_amount);
+CREATE INDEX IF NOT EXISTS view_convert_idx_2 ON bancor.view_convert (target_token_address) INCLUDE (target_token_amount);
+CREATE INDEX IF NOT EXISTS view_convert_idx_3 ON bancor.view_convert (block_time);
```
Comment (mewwts) [inline schema/bancor/view_convert.sql:80]: Non-materialized views don't have indexes, so these lines should be removed.

File: schema/bancor/view_smart_tokens.sql [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    ('XIOUSDB'::text, '\x18d8001d1da44fe96f442f5980e08d2ab4e19594'::bytea, 18::numeric),
+    ('XNKBNT'::text, '\x1b4d8c62ddf6947616a5fcda4ca40a8715d2a4cb'::bytea, 18::numeric),
+    ('XPATBNT'::text, '\xee769ce6b4e2c2a079c5f67081225af7c89f874c'::bytea, 18::numeric),
+    ('YHTSBNT'::text, '\x04a3030c94fb2dbe2b898d8cbf6fd1c656fa69dd'::bytea, 18::numeric),
+    ('ZIPTBNT'::text, '\xc4a01182ab1e502a1c1d17024e4924573ce001cc'::bytea, 18::numeric),
+    ('ZRXUSDB'::text, '\x1a3c6768e200482f5f47d1be77b7255abcae4fe2'::bytea, 18::numeric)
+;
+
+CREATE UNIQUE INDEX IF NOT EXISTS view_smart_tokens_id ON bancor.view_smart_tokens (contract_address);
+CREATE INDEX view_tokens_1 ON bancor.view_smart_tokens (symbol);
```
Comment (mewwts) [inline schema/bancor/view_smart_tokens.sql:274]: This view is non-materialized, so can't have indexes.

File: schema/bancor/view_smart_tokens.sql [hunk 3]
```diff
@@ -0,0 +1,274 @@
+CREATE OR REPLACE VIEW bancor.view_smart_tokens (symbol, contract_address, decimals) AS VALUES
```
Comment (mewwts) [inline schema/bancor/view_smart_tokens.sql:1]: Are there no events when smart-tokens are added, such that we could compute this dynamically?

File: schema/erc20/tokens.sql [hunk 4]
```diff
@@ -143,12 +181,46 @@ COPY erc20.tokens (contract_address, symbol, decimals) FROM stdin;
...
+\\xbfc615ab171d10a4e6498549cd3a46766feefe01	FNCD	18
+\\xc3308318eca1eab76d2f4b52c310289ae526c8ca	FNCD	18
+\\xce6437ee4932cfba3ecaefb93a1929a379b13b24	FNCD	18
+\\xcf97b4f961fa247cd60eb3cbf92a135fa2c9cc5d	FNCD	18
+\\xd2a3e62ca1658421a8a711bd1f8ed9d01780e6e6	FNCD	18
+\\xd609609fdf3a6ea89d5ea981af5ce5c3c3daa44a	FNCD	18
+\\xe23a3a78b131553c0bab4323bba7f61f6950bdaa	FNCD	18
+\\xf0fd2500bd31a56cf1a3839e32fd28e11cda60b3	FNCD	18
+\\xfb6f6467724b7cdc75a1977812d78b7467f69d77	FNCD	18
+\\xfc382c9f9f65a66bbd3b35e10e7ac5a7abf11c8f	FNCD	18
```
Comment (mewwts) [inline schema/erc20/tokens.sql:213]: Are these different versions of the same token or what's going on here?

Comment (milkyklim) [reply schema/erc20/tokens.sql:213]: Those are indeed different versions of the same token. Don't know how useful there are but there are in the list of tokens traded (at least once) on bancor.

File: schema/bancor/view_smart_tokens.sql (hunk 3 above)
Comment (milkyklim) [reply schema/bancor/view_smart_tokens.sql:1]: These are created in `BancorConverterRegistry_evt_SmartTokenAdded`. Problem, event doesn't contain `symbol` and `decimals`. If you have an idea how to fix this – glad to help.

For ref:
```
SELECT *
FROM bancor."BancorConverterRegistry_evt_SmartTokenAdded";
```

**Comment by milkyklim** [comment]
> Also, is the view_smart_tokens needed if all are added to erc20.tokens? Or perhaps you only added a subset to erc20.tokens?

@mewwts `erc20.tokens` contains normal tokens e.g. EOS or NMR while `view_smart_tokens` contains underlying bancor tokens, e.g. EOSBNT or NMRBNT. Does it make sense to you?

File: schema/bancor/view_smart_tokens.sql (hunk 3 above)
Comment (mewwts) [reply schema/bancor/view_smart_tokens.sql:1]: Got it! Yeah, for now this is OK!

File: schema/bancor/view_smart_tokens.sql [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
```
Comment (mewwts) [inline schema/bancor/view_smart_tokens.sql:1]: This should ideally match the order of `erc20.tokens` i.e. (contract_address, symbol, decimals)

File: schema/bancor/view_convert.sql [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+       conversion_fee / 10^t2.decimals * p2.price AS conversion_usd_fee,
+       s.contract_address,
+       tx_hash,
+       block_time
+FROM conversions s
+LEFT JOIN
+  (SELECT *
+   FROM erc20.tokens
+   UNION
+   SELECT *
```
Comment (mewwts) [inline schema/bancor/view_convert.sql:65]: See above comment, this union doesn't work unless the types are aligned.

File: schema/bancor/view_smart_tokens.sql (hunk 5 above)
Comment (milkyklim) [reply schema/bancor/view_smart_tokens.sql]: Sorry for that one, tricky to debug when can't really test the views. :) 
Committed order fix.

**Comment by mewwts** [comment]
Awesome! Thanks @milkyklim :+1:

**Comment by mewwts** [comment]
image
Live on dune!

**Pull request merged**